File: 22.GenerateParenthesis/Solution.java
import java.util.ArrayList;
import java.util.List;

public class Solution {
    private List<String> ans = new ArrayList<>();

    private void generate(int o, int c, int n, String s) {
        if (s.length() == 2 * n) {
            ans.add(s);
        }
        if (o < n)
            generate(o + 1, c, n, s + '(');
        if (c < o)
            generate(o, c + 1, n, s + ')');
    }

    public List<String> generateParenthesis(int n) {
        generate(0, 0, n, "");
        return ans;
    }
}

File: 22.GenerateParenthesis/Solution.py
from typing import List
class Solution:
    ans = []
    def recurse(self,open:int,close:int,n:int,s:str)->None:
        if len(s) == 2*n:
            self.ans.append(s)
            return
        if open < n: self.recurse(open + 1, close,n, s + '(')
        if close < open: self.recurse(open,close+ 1, n, s+ ')')
            
    def generateParenthesis(self,n:int)-> List[str]:
        self.ans.clear()
        self.recurse(0,0,n,"")
        return self.ans
        